### ISSUE TITLE
Make code relocatable (PIC) by default

### ISSUE THREAD

**Issue by webisu** [NOAA-EMC/NCEPLIBS-ip#261]
Compile with -fPIC to get relocatable code.  Want to make a (relocatable) python library that uses iplib.

**Comment by AlexanderRichert-NOAA** [comment]
Do you need this through spack-stack? If you just need it for your own development copy of ip, then building with `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` should do the trick (I assume you're compiling a static library).

**Comment by webisu** [comment]
The goal is to make python call libwgrib2.so (import pywgrib2_s).  This
means that libraries
that wgrib2 calls will have to be compiled with -fPIC.

**Comment by AlexanderRichert-NOAA** [comment]
Have you already tried `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` for wgrib2? If that doesn't work then that's probably a cmake bug for wgrib2.

**Comment by webisu** [comment]
I got all the netcdf, hdf5, g2c, png, jasper, openjpeg, aec libraries to
work.
IP was the only exception.  I can make wgrib2 when I turned off
MAKE_SHARED_LIB
in wgrib2.

**Comment by AlexanderRichert-NOAA** [comment]
To investigate further I'll need exact steps to reproduce the problem, please.

**Comment by webisu** [comment]
How I tell whether  *.a is relocatable:
 $ ar x libip_4.a
 $ gcc -shared -o libtest.so *.o
Complaints about -fPIC.

I built ip by:
git clone "https://github.com/NOAA-EMC/NCEPLIBS-ip"
cd  NCEPLIB-ip
edit CMakeLists.txt
   change option(OPENMP "Use OpenMP threading" OFF)
   to     option(OPENMP "Use OpenMP threading" ON)
mkdir build
cd build
make
sudo make install

**Comment by AlexanderRichert-NOAA** [comment]
I don't see the cmake step-- are you using `-DCMAKE_POSITION_INDEPENDENT_CODE=ON`? i.e., `cmake .. -DCMAKE_POSITION_INDEPENDENT_CODE=ON`?

**Comment by webisu** [comment]
No.  So I have to make a non-default version of IP.

**Comment by AlexanderRichert-NOAA** [comment]
Correct, it'll require either explicit PIC with the above build option, or building the shared version.

**Comment by webisu** [comment]
Of course it works if you rebuild the IP package
with -DCMAKE_POSITION_INDEPENDENT_CODE=ON.
The point is that the default cmake options does not produce a library that
is usable in some
applications.  One could argue that the position independent code is a
unusual feature and
is not worth the performance penalty.  Anyways I need to know the IPlib
policy for future wgrib2
documentation.

**Comment by AlexanderRichert-NOAA** [comment]
I'd be open to changing the default if it doesn't have a significant impact on performance. In the case of, say, g2c, I would have expected that it would have the same issue. Which version of ip are you using, is this the head of develop?

**Comment by webisu** [comment]
Performance is important but people tend not to worry about ip and g2/g2c
because
those are used in I/O bound processes (at least for the models).  The time
sensitive
NCEP uses of wgrib2 are NOMADS and the gfs post-processing.  The NOMADS job
doesn't
use IP.  The post-processing uses IP but tends to be I/O bound.

G2c is position independent, and I am using the develop version of IP.
Openblas,
PNG, AEC libraries from Ubuntu were position independent. The Jasper build
was by default position independent.

**Comment by edwardhartnett** [comment]
@webisu needs a release of ip after this feature, in support of some operational code with wgrib2. So can we get a new release?

**Comment by AlexanderRichert-NOAA** [comment]
Yes, I'll do so shortly

**Comment by AlexanderRichert-NOAA** [comment]
https://github.com/NOAA-EMC/NCEPLIBS-ip/releases/tag/v5.2.0
